src/history/hs.c

```diff
@@ -512,6 +491,52 @@ __hs_insert_record_with_btree(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BT
     while ((ret = __hs_insert_record_with_btree_int(
               session, cursor, btree, key, upd, type, hs_value, stop_time_point)) == WT_RESTART)
         ;
+    WT_ERR(ret);
+
+    /* If we inserted a timestamped update, we don't need to delete any history store records. */
+    if (upd->start_ts != WT_TS_NONE)
+        goto done;
+
+/*
+ * If we inserted an update with no timestamp, we need to delete all history records for that key
+ * that are further in the history table than us (the key is lexicographically greater). For
+ * timestamped tables that are occasionally getting a non-timestamped update, that means that all
+ * timestamped updates should get removed. In the case of non-timestamped tables, that means that
+ * all updates with higher transaction ids will get removed (which could happen at some more relaxed
+ * isolation levels).
+ */
+#ifdef HAVE_DIAGNOSTIC
+    /*
+     * We need to initialize the last searched key so that we can do key comparisons when we
+     * begin iterating over the history store. This needs to be done otherwise the subsequent
+     * "next" calls will blow up.
+     */
+    WT_ERR(__wt_cursor_key_order_init((WT_CURSOR_BTREE *)cursor));
+#endif
+    F_SET(cursor, WT_CURSTD_IGNORE_TOMBSTONE);
+
+    /* We're pointing at the newly inserted update. Iterate once more to avoid deleting it. */
+    while ((ret = cursor->next(cursor)) == WT_RESTART)
+        ;
```

Comment:
I don't think WT_RESTART will ever be returned out of a `cursor->next` call - the call should handle retrying on WT_RESTART internally.

Reply:
5d76f2c